scrape/scrape_test.go

```diff
@@ -4757,70 +4754,34 @@ metric: <
-					MetricNameEscapingScheme:       model.AllowUTF8,
-				}
-
-				scrapeCount := 0
-				scraped := make(chan bool)
-
-				ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, _ *http.Request) {
-					if metricsText.contentType != "" {
-						w.Header().Set("Content-Type", `application/vnd.google.protobuf; proto=io.prometheus.client.MetricFamily; encoding=delimited`)
-						for _, text := range metricsText.text {
-							buf := &bytes.Buffer{}
-							// In case of protobuf, we have to create the binary representation.
-							pb := &dto.MetricFamily{}
-							// From text to proto message.
-							require.NoError(t, proto.UnmarshalText(text, pb))
-							// From proto message to binary protobuf.
-							protoBuf, err := proto.Marshal(pb)
-							require.NoError(t, err)
-
-							// Write first length, then binary protobuf.
-							varintBuf := binary.AppendUvarint(nil, uint64(len(protoBuf)))
-							buf.Write(varintBuf)
-							buf.Write(protoBuf)
-							w.Write(buf.Bytes())
-						}
-					} else {
-						for _, text := range metricsText.text {
-							fmt.Fprint(w, text)
-						}
-					}
-					scrapeCount++
-					if scrapeCount > 2 {
-						close(scraped)
+				sl := newBasicScrapeLoop(t, context.Background(), nil, func(ctx context.Context) storage.Appender { return simpleStorage.Appender(ctx) }, 0)
+				sl.alwaysScrapeClassicHist = tc.alwaysScrapeClassicHistograms
+				sl.convertClassicHistToNHCB = tc.convertClassicHistToNHCB
+				sl.enableNativeHistogramIngestion = true
+				app := simpleStorage.Appender(context.Background())
+
+				var content []byte
+				contentType := metricsText.contentType
+				switch contentType {
+				case "application/vnd.google.protobuf":
+					buf := &bytes.Buffer{}
+					for _, text := range metricsText.text {
+						// In case of protobuf, we have to create the binary representation.
+						pb := &dto.MetricFamily{}
+						// From text to proto message.
+						require.NoError(t, proto.UnmarshalText(text, pb))
+						// From proto message to binary protobuf.
+						buf.Write(protoMarshalDelimited(t, pb))
 					}
-				}))
-				defer ts.Close()
-
-				sp, err := newScrapePool(config, simpleStorage, 0, nil, nil, &Options{DiscoveryReloadInterval: model.Duration(10 * time.Millisecond), EnableNativeHistogramsIngestion: true}, newTestScrapeMetrics(t))
-				require.NoError(t, err)
-				defer sp.stop()
-
-				testURL, err := url.Parse(ts.URL)
-				require.NoError(t, err)
-				sp.Sync([]*targetgroup.Group{
-					{
-						Targets: []model.LabelSet{{model.AddressLabel: model.LabelValue(testURL.Host)}},
-					},
-				})
-				require.Eventually(t, func() bool { return len(sp.ActiveTargets()) == 1 }, 5*time.Second, 50*time.Millisecond)
-
-				select {
-				case <-time.After(5 * time.Second):
-					t.Fatalf("target was not scraped")
-				case <-scraped:
+					content = buf.Bytes()
+				case "text/plain", "":
+					content = []byte(strings.Join(metricsText.text, ""))
```

Comment:
I had to hunt around to figure out what this `strings.Join` is doing.
I think the crux is that `genTestCounterText` creates a string with a newline at the end, so you can just concatenate them all together.
Which is OK, but maybe a couple of comments noting these constraints would help?